[scheduler-extender] Add common scheduler extender; [agent] use direct LVM queries and fix BlockDeviceFilter handling

krpsh123 opened the pull request:
## Description

Move the scheduler extender from `sds-local-volume` to `sds-node-configurator` and introduce a common extender capable of scheduling both local and replicated volumes from a single component.

Additionally, fix two independent issues in the agent:
- avoid stale LV reads after LVM mutations by calling `GetLV` directly instead of relying on the cache;
- prevent `BlockDeviceFilter` crash when `In`/`NotIn` match expressions have nil or empty values.

This change affects scheduling-related components and agent behavior. After rollout, the new common scheduler extender and its TLS resources will be deployed in the cluster.

## Why do we need it, and what problem does it solve?

Previously, the scheduler extender lived in `sds-local-volume`, making it unsuitable as a shared scheduling component for different SDS volume types. This change consolidates scheduling in `sds-node-configurator` and makes the extender aware of both local and replicated storage.

The agent fixes address two production-observable issues:
- after `lvcreate`/`lvextend`, the in-memory cache contained stale data until the next scanner cycle, causing unnecessary 5 s requeues and busy-wait loops;
- user-created `BlockDeviceFilter` resources with `In`/`NotIn` and empty `values` caused `LabelSelectorAsSelector` to fail, breaking the entire block device reconciliation loop.

## What is the expected result?

After applying the changes:

- the common scheduler extender is deployed from `sds-node-configurator` with TLS assets and correct RBAC;
- pods that use supported PVCs are filtered, scored, and scheduled by the common extender for both local and replicated volumes;
- replicated storage classes and pools are resolved via `GetStoragePoolName()` matching the current CRD layout;
- agent reconciliation reads LV data directly from LVM after mutations, eliminating stale-cache delays;
- `BlockDeviceFilter` processing gracefully skips `In`/`NotIn` expressions with empty values instead of crashing.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

krpsh123 committed:
[scheduler-extender] Add common scheduler extender for local and replicated volumes

Move the scheduler extender from sds-local-volume to sds-node-configurator
and introduce a common extender that can serve multiple SDS volume types.

- add the common scheduler extender image, deployment, and TLS wiring
- integrate replicated storage support and align RSC/RSP handling with the current CRD
- refactor cache to reservation store with lazy TTL and stored offset for race safety
- add bind endpoint, non-Ready LVG filtering, request-scoped contexts, and /stat counters
- add scheduler tests and linter cleanup

Signed-off-by: Pavel Karpov <pavel.karpov@flant.com>
Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin committed:
[agent] Use GetLV instead of cache after LVM mutations to avoid delays

After lvcreate and lvextend, the in-memory cache contains stale data
until the scanner runs. This caused unnecessary 5s requeues and
blocking busy-wait loops.

- LLV create: replace getLVActualSize with commands.GetLV after lvcreate
- LLV resize: replace getLVActualSize with commands.GetLV after lvextend
- LLV extender: replace FindLV busy-wait loop with GetLV after lvextend
- LLVS snapshot: use GetLV after CreateThinLogicalVolumeSnapshot instead
  of requeueing for cache discovery

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin committed:
[agent] Fix BlockDeviceFilter crash on In/NotIn with empty values

BlockDeviceFilter resources with In/NotIn matchExpressions and nil/empty
values caused metav1.LabelSelectorAsSelector to fail, breaking the
entire block device reconciliation loop.

Add sanitizeLabelSelector() that drops such vacuous expressions before
parsing. Add tests for nil values, empty values, and mixed cases.

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin committed:
[scheduler-extender, agent] Fix review findings

- Fix SC-to-PVC validation: replace incorrect length comparison with
  per-PVC lookup (a Pod with two PVCs sharing the same SC triggered
  a false 500). Add handler-level tests for filter and prioritize.
- Return errors from extractRequestedSize instead of silently skipping
  PVCs with continue.
- Optimize getStorageClassesUsedByPVCs: Get only referenced SCs
  instead of listing all in the cluster.
- Remove incorrect PhaseFailed in llv-extender for transient cases
  (stale cache, GetLV after successful extend) — just retry.

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin commented on images/sds-common-scheduler-extender/pkg/scheduler/func.go:
```diff
@@ -0,0 +1,657 @@
+	log logger.Logger,
+	pvcs map[string]*corev1.PersistentVolumeClaim,
+	scs map[string]*storagev1.StorageClass,
+) (map[string]PVCRequest, error) {
```
Agreed, fixed. Replaced continue with return error — it's better to fail explicitly than to make scheduling decisions based on incomplete storage requirements.

AleksZimin commented on images/agent/internal/controller/llv_extender/reconciler.go:
```diff
@@ -225,32 +225,23 @@ func (r *Reconciler) ReconcileLVMLogicalVolumeExtension(
 			shouldRetry = true
-
-			if err = r.llvCl.UpdatePhaseIfNeeded(ctx, &llv, v1alpha1.PhaseFailed, err.Error()); err != nil {
+			if err = r.llvCl.UpdatePhaseIfNeeded(ctx, &llv, v1alpha1.PhaseFailed, getLVErr.Error()); err != nil {
```
Good point. Removed PhaseFailed for this case — the extension already succeeded, only the verification via GetLV failed. Keeping Resizing phase and retrying is more accurate. Also removed PhaseFailed for the "LV not found in cache" case (same reasoning — cache staleness is transient).

AleksZimin committed:
[scheduler-extender] Return empty node list instead of 500 on scheduling-critical errors

With failurePolicy: Ignore, returning 500 causes the scheduler to
silently skip the extender. For errors where we cannot safely determine
storage placement (missing SC, failed size extraction, etc.), return
HTTP 200 with an empty NodeNames and all nodes in FailedNodes. This
makes the Pod stay Pending instead of being scheduled blindly.

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin committed:
[scheduler-extender] Extend local PVC reservation TTL to 10 minutes
Under heavy scheduling, CSI could take longer than 60s to create LLV CRs.
Reservations expired before that, so GetReservedSpace stopped counting
pending allocations while sumLLVSpace still did not see new LLVs, which
led to over-scheduling and LV creation failures on thin pools.
- defaultReservationTTL: 60s -> 600s (filter)
- defaultNarrowTTL: 60s -> 600s (PVC watcher on selected-node)
Reservations are still cleared on Bound PVC / LLV Created; TTL remains a
safety net for orphaned entries.

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin committed:
Revert "[scheduler-extender] Extend local PVC reservation TTL to 10 minutes"

This reverts commit c92e3952de5fd2696c9ecff258fab1f5323b5cae.

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin committed:
Merge branch 'main' into add-common-scheduler-extender-v2

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

AleksZimin commented on images/sds-common-scheduler-extender/pkg/scheduler/prioritize.go:
```diff
@@ -0,0 +1,402 @@
+	"github.com/deckhouse/sds-node-configurator/images/sds-common-scheduler-extender/pkg/logger"
+)
+
+func (s *scheduler) prioritize(w http.ResponseWriter, r *http.Request) {
```
prioritize can't filter nodes, it can only set scores. There is no fail-closed equivalent for scoring: whether we return 500 (ignored by failurePolicy: Ignore) or return all-zero scores, the result is the same — the scheduler proceeds without storage-aware scoring from this extender.
